File: wasteservice.core/test/it/unibo/lenziguerra/wasteservice/TestDeposit.java
package it.unibo.lenziguerra.wasteservice;

import it.unibo.kactor.*;
import it.unibo.lenziguerra.wasteservice.utils.PrologUtils;
import org.apache.log4j.Level;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import unibo.actor22comm.coap.CoapConnection;
import unibo.actor22comm.utils.ColorsOut;
import unibo.actor22comm.utils.CommSystemConfig;
import unibo.actor22comm.utils.CommUtils;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static org.junit.Assert.*;

public class TestDeposit {
    private String ctx_wasteservice;
    private String ctx_trolley;
    private String ctx_storage;

    TrolleyPosObserver trolleyPosObserver;
    WasteServiceTrolleyPosObserver wasteServiceObserver;

    Integer[][] homeCoords;
    Integer[][] indoorCoords;
    Integer[][] glassBoxCoords;
    Integer[][] plasticBoxCoords;

    @Before
    public void up() {
        CommSystemConfig.tracing = false;

        SystemConfig.INSTANCE.setConfiguration("SystemConfig.json", false);
        SystemConfig.INSTANCE.getPositions().put("home", List.of(List.of(0, 0),List.of(0, 0)));
        SystemConfig.INSTANCE.getPositions().put("indoor", List.of(List.of(0, 3),List.of(1, 3)));
        SystemConfig.INSTANCE.getPositions().put("plastic_box", List.of(List.of(2, 0),List.of(3, 0)));
        SystemConfig.INSTANCE.getPositions().put("glass_box", List.of(List.of(3, 2),List.of(3, 3)));

        configCoords();

        new Thread(() -> {
            new RunCtxTestDeposit().main();
        }).start();

        waitForActors();
        startTrolleyCoapConnection();
        startWasteServiceCoapConnection();
    }

    @After
    public void down() {
        ColorsOut.outappl(this.getClass().getName() + " TEST END", ColorsOut.GREEN);
    }

    @Test
    public void testPositionsGlass() {
        startDeposit("glass", 10);
        List<String> expectedPositions = List.of("home", "indoor", "glass_box");
        List<Integer[][]> expectedPositionsCoords = List.of(homeCoords, indoorCoords, glassBoxCoords);

        int maxSecondsWait = 10;
        positionsTest(expectedPositions, expectedPositionsCoords, maxSecondsWait);
    }

    @Test
    public void testPositionsPlastic() {
        startDeposit("plastic", 10);
        List<String> expectedPositions = List.of("home", "indoor", "plastic_box");
        List<Integer[][]> expectedPositionsCoords = List.of(homeCoords, indoorCoords, plasticBoxCoords);

        int maxSecondsWait = 10;
        positionsTest(expectedPositions, expectedPositionsCoords, maxSecondsWait);
    }

    @Test
    public void testDeposit() {
        dispatch("storage", MsgUtil.buildDispatch(
            "test",
            "testStorageReset",
            "testStorageReset()",
                SystemConfig.INSTANCE.getCtxNames().get("storage")
        ));

        startDeposit("glass", 15);

        int maxSecondsWait = 10;
        for (int i = 0; i < maxSecondsWait; i++) {
            CommUtils.delay(1000);
            String lastPos = wasteServiceObserver.getHistory().get(wasteServiceObserver.getHistory().size() - 1);
            if (lastPos.equals("error")) {
                fail("Movement error! Pos history is " + wasteServiceObserver.getHistory());
            }
            else if (lastPos.equals("glass_box")) {
                // Lascia tempo di scaricare pesi nel caso la posizione
                // sia stata appena raggiunta
                CommUtils.delay(1000);

                List<String> storageReplyLines = PrologUtils.INSTANCE.getFuncLines(
                        coapRequest("storage", ctx_storage, SystemConfig.INSTANCE.getCtxNames().get("storage")),
                        "content"
                );
                for (String line : storageReplyLines) {
                    List<String> args =  PrologUtils.INSTANCE.extractPayload(line);
                    float value = Float.parseFloat(args.get(1));
                    switch (args.get(0)) {
                        case "glass": assertEquals(15f, value, 0.0001f); break;
                        case "plastic": assertEquals(0f, value, 0.0001f); break;
                    }
                }
                return;
            }
        }
        fail("Too much time to reach final position, pos history was:<" + trolleyPosObserver.getHistory() + ">");
    }


    protected void positionsTest(List<String> expectedPositions, List<Integer[][]>  expectedCoords, int maxSecondsWait) {
        long startTime = System.currentTimeMillis();
        while (System.currentTimeMillis() - startTime < maxSecondsWait) {
            waitForObserverUpdate(1000);

            List<String> posHistory = wasteServiceObserver.getHistory();
            if (posHistory.get(posHistory.size() - 1).equals("error")) {
                fail("Movement error! Pos history is " + posHistory);
            }
            for (int j = 0; j < expectedPositions.size(); j++) {
                if (j >= posHistory.size()) {
                    break;
                } else {
                    assertEquals(expectedPositions.get(j), posHistory.get(j));
                }
            }

            // Se ogni posizione corrisponde, e si è raggiunta l'ultima, ritorna con successo
            // se corrispondono anche le coordinate
            if (expectedPositions.size() == posHistory.size()) {
                // Controlla che passi dalle coordinate richieste (non SOLO le coordinate richieste)
                List<int[]> coordHistory = trolleyPosObserver.getHistory();
                int matchedCoords = 0;
                for (int[] coord : coordHistory) {
                    if (matchesOneCoordInRectangle(coord, expectedCoords.get(matchedCoords))) {
                        matchedCoords++;
                    }
                }
                if (matchedCoords == expectedCoords.size()) {
                    return;
                } else {
                    fail("Finished positions but coordinates didn't match: "
                            + coordHistory.stream().map(Arrays::toString).collect(Collectors.joining(","))
                    );
                }
            }
        }
        fail("Too much time to reach final position, pos history was:<" + trolleyPosObserver.getHistory() + ">");
    }

    private void waitForObserverUpdate(int maxTimeMillis) {
        try {
            if (!wasteServiceObserver.getSemaphore().tryAcquire(1, maxTimeMillis, TimeUnit.MILLISECONDS)) {
                ColorsOut.outappl("Position check timeout", ColorsOut.BLUE);
            }
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    protected boolean matchesOneCoordInRectangle(int[] coord, Integer[][] checkCorners) {
        for (int i = checkCorners[0][0]; i <= checkCorners[1][0]; i++) {
            for (int j = checkCorners[0][1]; j <= checkCorners[1][1]; j++) {
                if (coord[0] == i && coord[1] == j)
                    return true;
            }
        }
        return false;
    }

    protected void startDeposit(String type, int amount) {
        String startDepositDispatch = MsgUtil.buildRequest("test", "triggerDeposit",
                "triggerDeposit(" + type + ", " + amount + ")",
                SystemConfig.INSTANCE.getCtxNames().get("wasteService")
        ).toString();
        try {
            ConnTcp connTcp = new ConnTcp(
                    SystemConfig.INSTANCE.getHosts().get("wasteServiceContext"),
                    SystemConfig.INSTANCE.getPorts().get("wasteServiceContext")
            );
            connTcp.request(startDepositDispatch);
            ColorsOut.outappl("STARTED DEPOSIT VIA DISPATCH", ColorsOut.GREEN);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    protected void dispatch(String hostname, IApplMessage msg) {
        try {
            ConnTcp connTcp = new ConnTcp(
                    SystemConfig.INSTANCE.getHosts().get(hostname),
                    SystemConfig.INSTANCE.getPorts().get(hostname)
            );
            connTcp.forward(msg.toString());
        } catch (Exception e) {
            e.printStackTrace();
            fail(e.getMessage());
        }
    }

    protected void waitForActors() {
        ColorsOut.outappl(this.getClass().getName() + " waits for actors ... ", ColorsOut.GREEN);
        ActorBasic trolley = QakContext.Companion.getActor(SystemConfig.INSTANCE.getCtxNames().get("trolley"));
        while (trolley == null) {
            CommUtils.delay(200);
            trolley = QakContext.Companion.getActor(SystemConfig.INSTANCE.getCtxNames().get("trolley"));
        }
        ActorBasic wasteservice = QakContext.Companion.getActor(SystemConfig.INSTANCE.getCtxNames().get("wasteService"));
        while (wasteservice == null) {
            CommUtils.delay(200);
            wasteservice = QakContext.Companion.getActor(SystemConfig.INSTANCE.getCtxNames().get("wasteService"));
        }
        ActorBasic storage = QakContext.Companion.getActor(SystemConfig.INSTANCE.getCtxNames().get("storage"));
        while (storage == null) {
            CommUtils.delay(200);
            storage = QakContext.Companion.getActor(SystemConfig.INSTANCE.getCtxNames().get("storage"));
        }

        ctx_trolley = trolley.getContext().getName();
        ctx_wasteservice = wasteservice.getContext().getName();
        ctx_storage = storage.getContext().getName();

        ColorsOut.outappl(String.format("Actors loaded, contexts are: %s, %s, %s", ctx_trolley, ctx_wasteservice, ctx_storage), ColorsOut.GREEN);
    }

    protected void startTrolleyCoapConnection() {
        trolleyPosObserver = new TrolleyPosObserver();
        new Thread(() -> {
            CoapConnection conn = new CoapConnection(SystemConfig.INSTANCE.getHosts().get("trolley")
                    + ":" + SystemConfig.INSTANCE.getPorts().get("trolley"),
                    ctx_trolley + "/" + SystemConfig.INSTANCE.getCtxNames().get("trolley")
            );
            conn.observeResource(trolleyPosObserver);
            ColorsOut.outappl("connected via Coap conn:" + conn , ColorsOut.CYAN);
        }).start();
    }

    protected void startWasteServiceCoapConnection() {
        wasteServiceObserver = new WasteServiceTrolleyPosObserver();
        new Thread(() -> {
            CoapConnection conn = new CoapConnection(SystemConfig.INSTANCE.getHosts().get("wasteServiceContext")
                    + ":" + SystemConfig.INSTANCE.getPorts().get("wasteServiceContext"),
                    ctx_wasteservice + "/" + SystemConfig.INSTANCE.getCtxNames().get("wasteService")
            );
            conn.observeResource(wasteServiceObserver);
            ColorsOut.outappl("connected via Coap conn:" + conn , ColorsOut.CYAN);
        }).start();
    }

    protected String coapRequest(String hostname, String context, String actor){
        CoapConnection reqConn = new CoapConnection(
                SystemConfig.INSTANCE.getHosts().get(hostname) + ":"
                        + SystemConfig.INSTANCE.getPorts().get(hostname),
                context + "/" + actor
        );
        String answer = reqConn.request("");
        ColorsOut.outappl("coapRequest answer=" + answer, ColorsOut.CYAN);
        return answer;
    }

    private Integer[][] doubleListToArray(List<List<Integer>> list) {
        return list.stream().map(l -> l.toArray(new Integer[0]))
                .toArray(Integer[][]::new);
    }

    protected void configCoords() {
        // Carica da config in futuro
        Map<String, List<List<Integer>>> positions = SystemConfig.INSTANCE.getPositions();
        homeCoords = doubleListToArray(positions.get("home"));
        indoorCoords = doubleListToArray(positions.get("indoor"));
        plasticBoxCoords = doubleListToArray(positions.get("plastic_box"));
        glassBoxCoords = doubleListToArray(positions.get("glass_box"));
    }
}

File: wasteservice.core/test/it/unibo/lenziguerra/wasteservice/TestDeposit.kt
package it.unibo.lenziguerra.wasteservice

import it.unibo.kactor.MsgUtil
import it.unibo.kactor.MsgUtil.buildDispatch
import it.unibo.kactor.MsgUtil.buildRequest
import it.unibo.kactor.QakContext
import it.unibo.kactor.sysUtil
import it.unibo.lenziguerra.wasteservice.SystemConfig.contexts
import it.unibo.lenziguerra.wasteservice.data.StorageStatus
import it.unibo.lenziguerra.wasteservice.utils.LogUtils
import it.unibo.lenziguerra.wasteservice.utils.MsgUtilsWs
import it.unibo.lenziguerra.wasteservice.utils.PrologUtils.extractPayload
import it.unibo.lenziguerra.wasteservice.utils.PrologUtils.getFuncLines
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions
import org.eclipse.californium.core.CoapObserveRelation
import org.junit.Assert
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.assertEquals
import unibo.comm22.coap.CoapConnection
import unibo.comm22.utils.ColorsOut
import unibo.comm22.utils.CommUtils
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread


class TestDeposit {
    companion object {
        const val TEST_CONTEXT_NAME = "ctx_wasteservice_test"
        const val TEST_CONTEXT_HOST = "localhost"
        const val TEST_CONTEXT_PORT = 9651
        const val TEST_CONTEXT_DESC = """context($TEST_CONTEXT_NAME, "$TEST_CONTEXT_HOST",  "TCP", "$TEST_CONTEXT_PORT").
            qactor( storagemanager, $TEST_CONTEXT_NAME, "it.unibo.storagemanager.Storagemanager").
            qactor( wasteservice, $TEST_CONTEXT_NAME, "it.unibo.wasteservice.Wasteservice").
            qactor( trolley, $TEST_CONTEXT_NAME, "it.unibo.trolley.Trolley").
        """

        lateinit var qakContext: QakContext

        lateinit var trolleyCoapConnection: CoapConnection
        lateinit var wasteserviceCoapConnection: CoapConnection

        @BeforeAll
        @JvmStatic
        fun upClass() {
            SystemConfig.positions["home"] = listOf(listOf(0, 0), listOf(0, 0))
            SystemConfig.positions["indoor"] = listOf(listOf(0, 1), listOf(1, 1))
            SystemConfig.positions["plastic_box"] = listOf(listOf(2, 1), listOf(2, 1))
            SystemConfig.positions["glass_box"] = listOf(listOf(3, 1), listOf(3, 1))

            SystemConfig.hosts.replaceAll { _, _ -> "localhost" }
            SystemConfig.contexts.forEach {
                SystemConfig.ports[it.key] = TEST_CONTEXT_PORT
            }
            SystemConfig.contexts.replaceAll { _, _ -> TEST_CONTEXT_NAME }

            SystemConfig.disableRead()

            thread { runBlocking {
                ContextTestUtils.createContextsFromString("localhost", this, TEST_CONTEXT_DESC, "sysRules.pl")
            } }

            waitForActors()
            addTestActors()
            startCoapConnections()

            LogUtils.threadOut("TestDeposit ready!", ColorsOut.GREEN)
        }

        @AfterAll
        @JvmStatic
        fun downClass() {
            qakContext.terminateTheContext()
        }

        fun waitForActors() {
            val actorsToWait = listOf("storagemanager", "wasteservice", "trolley")

            LogUtils.threadOut(this::class.java.name + " waits for actors ... ", ColorsOut.GREEN)

            actorsToWait.forEach {
                var actor = QakContext.getActor(it)
                while (actor == null) {
                    CommUtils.delay(200)
                    actor = QakContext.getActor(it)
                }
            }

            qakContext = sysUtil.getContext(TEST_CONTEXT_NAME)!!

            LogUtils.threadOut("Actors loaded", ColorsOut.GREEN)
        }

        fun addTestActors() {
            val pathexec = PathExecDummyImmediate("pathexecstop")
            qakContext.addActor(pathexec)

            LogUtils.threadOut("Added test actors <${pathexec.name} as ${pathexec::class}>", ColorsOut.GREEN)
        }

        fun startCoapConnections() {
            trolleyCoapConnection = CoapConnection(
                "$TEST_CONTEXT_HOST:$TEST_CONTEXT_PORT",
                "$TEST_CONTEXT_NAME/trolley"
            )
            LogUtils.threadOut("connected via Coap conn: $trolleyCoapConnection", ColorsOut.CYAN)
            wasteserviceCoapConnection = CoapConnection(
                "$TEST_CONTEXT_HOST:$TEST_CONTEXT_PORT",
                "$TEST_CONTEXT_NAME/wasteservice"
            )
            LogUtils.threadOut("connected via Coap conn: $wasteserviceCoapConnection", ColorsOut.CYAN)
        }
    }

    lateinit var trolleyPosObserver: TrolleyPosObserver
    lateinit var wasteServiceObserver: WasteServiceTrolleyPosObserver
    val coapObservers = mutableMapOf<CoapConnection, CoapObserveRelation>()

    @BeforeEach
    fun up() {
        startObservingTrolley()
        startObservingWasteservice()
        waitForHome()
    }

    @AfterEach
    fun down() {
        cleanupObservers()
        LogUtils.threadOut(this.javaClass.name + " TEST END", ColorsOut.GREEN)
    }

    @Test
    fun testPositionsGlass() {
        startDeposit("glass", 10)
        val expectedPositions = listOf("home", "indoor", "glass_box")
        val expectedPositionsCoords = listOf(
            SystemConfig.positions["home"]!!,
            SystemConfig.positions["indoor"]!!,
            SystemConfig.positions["glass_box"]!!,
        )
        val maxSecondsWait = 10
        positionsTest(expectedPositions, expectedPositionsCoords, maxSecondsWait)
    }

    @Test
    fun testPositionsPlastic() {
        startDeposit("plastic", 10)
        val expectedPositions = listOf("home", "indoor", "plastic_box")
        val expectedPositionsCoords = listOf(
            SystemConfig.positions["home"]!!,
            SystemConfig.positions["indoor"]!!,
            SystemConfig.positions["plastic_box"]!!,
        )
        val maxSecondsWait = 10
        positionsTest(expectedPositions, expectedPositionsCoords, maxSecondsWait)
    }

    @Test
    fun testDeposit() {
        dispatch(
            "testStorageReset",
            "testStorageReset()",
            SystemConfig.actors["storage"]!!
        )
        startDeposit("glass", 15)
        val maxSecondsWait = 10
        for (i in 0 until maxSecondsWait) {
            CommUtils.delay(1000)
            val lastPos = wasteServiceObserver.history.last()
            if (lastPos == "error") {
                fail("Movement error! Pos history is " + wasteServiceObserver.history)
            } else if (lastPos == "glass_box") {
                // Lascia tempo di scaricare pesi nel caso la posizione
                // sia stata appena raggiunta
                CommUtils.delay(1000)
                val storageStatus = StorageStatus.fromProlog(coapRequest("storage"))
                assertEquals(15.0f, storageStatus.amounts[WasteType.GLASS]!!, 0.001f)
                assertEquals(0f, storageStatus.amounts[WasteType.PLASTIC]!!, 0.001f)
                return
            }
        }
        fail("Too much time to reach final position, pos history was: ${wasteServiceObserver.history}")
    }

    private fun positionsTest(expectedPositions: List<String>, expectedCoords: List<List<List<Int>>>, maxSecondsWait: Int) {
        resetObserverSemaphores()
        val startTime = System.currentTimeMillis()
        while (System.currentTimeMillis() - startTime < maxSecondsWait * 1000) {
            waitForObserverUpdate(maxSecondsWait * 1000)
            val posHistory = wasteServiceObserver.history
            val coordHistory = trolleyPosObserver.history
            LogUtils.threadOut("positionsTest: start iteration, known histories: $posHistory, ${coordHistory.map { it.contentToString() }}", ColorsOut.CYAN)

            if (posHistory.last() == "error") {
                fail("Movement error! Pos history is " + posHistory);
            }
            for (j in expectedPositions.indices) {
                if (j >= posHistory.size) {
                    break
                } else {
                    assertEquals(expectedPositions[j], posHistory[j]);
                }
            }

            // Se ogni posizione corrisponde, e si è raggiunta l'ultima, ritorna con successo
            // se corrispondono anche le coordinate
            if (expectedPositions.size == posHistory.size) {
                // Controlla che passi dalle coordinate richieste (non SOLO le coordinate richieste)
                var matchedCoords = 0
                for (coord in coordHistory) {
                    if (matchesOneCoordInRectangle(coord, expectedCoords[matchedCoords])) {
                        matchedCoords++;
                    }
                }
                if (matchedCoords == expectedCoords.size) {
                    return;
                } else {
                    fail("Finished positions but coordinates didn't match: ${coordHistory.map { it.contentToString() }}")
                }
            }
        }
    }

    private fun resetObserverSemaphores() {
        wasteServiceObserver.semaphore.drainPermits()
        trolleyPosObserver.semaphore.drainPermits()
        LogUtils.threadOut("Reset semaphores for observers")
    }

    private fun waitForObserverUpdate(maxTimeMillis: Int) {
        try {
            if (!wasteServiceObserver.semaphore.tryAcquire(1, maxTimeMillis.toLong(), TimeUnit.MILLISECONDS)) {
                LogUtils.threadOut("Position check timeout", ColorsOut.BLUE)
            }
            if (!trolleyPosObserver.semaphore.tryAcquire(1, maxTimeMillis.toLong(), TimeUnit.MILLISECONDS)) {
                LogUtils.threadOut("Position check timeout", ColorsOut.BLUE)
            }
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }

    fun matchesOneCoordInRectangle(coord: IntArray, checkCorners: List<List<Int>>): Boolean {
        for (i in checkCorners[0][0]..checkCorners[1][0]) {
            for (j in checkCorners[0][1]..checkCorners[1][1]) {
                if (coord[0] == i && coord[1] == j) return true
            }
        }
        return false
    }


    fun startDeposit(type: String, amount: Int) {
        val startDepositDispatch = buildRequest(
            "test", "triggerDeposit",
            "triggerDeposit($type, $amount)",
            "wasteservice"
        ).toString()
        try {
            val connTcp = ConnTcp(TEST_CONTEXT_HOST, TEST_CONTEXT_PORT)
            connTcp.request(startDepositDispatch)
            LogUtils.threadOut("STARTED DEPOSIT VIA DISPATCH", ColorsOut.GREEN)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun coapRequest(element: String): String {
        val reqConn = CoapConnection(
            "${TEST_CONTEXT_HOST}:${TEST_CONTEXT_PORT}",
            "${TEST_CONTEXT_NAME}/${SystemConfig.actors[element]!!}"
        )
        val answer = reqConn.request("")
        LogUtils.threadOut("coapRequest answer=$answer", ColorsOut.CYAN)
        return if (answer != "0") {
            answer
        } else {
            throw IOException("Response null! Wrong element? Called with $element to ip " +
                    "${TEST_CONTEXT_HOST}:${TEST_CONTEXT_PORT}" +
                    "/${TEST_CONTEXT_NAME}/${SystemConfig.actors[element]!!}")
        }
    }

    fun dispatch(msgId: String, content: String, dest: String, actor: String = "test") {
        try {
            val msg = MsgUtil.buildDispatch(actor, msgId, content, dest)
            LogUtils.threadOut("Sending " + MsgUtilsWs.cleanMessage(msg), ColorsOut.CYAN)
            runBlocking {
                MsgUtil.sendMsg(msg, QakContext.getActor(msg.msgReceiver())!!)
            }
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
            Assertions.fail(e.message)
        }
    }

    fun startObservingTrolley() {
        trolleyPosObserver = TrolleyPosObserver()
        coapObservers[trolleyCoapConnection] = trolleyCoapConnection.observeResource(trolleyPosObserver)
        LogUtils.threadOut("Added trolley observer", ColorsOut.CYAN)

        // Force initial update
        CommUtils.delay(200)
        QakContext.getActor("wasteservice")!!.changed()
    }

    fun startObservingWasteservice() {
        wasteServiceObserver = WasteServiceTrolleyPosObserver()
        coapObservers[wasteserviceCoapConnection] = wasteserviceCoapConnection.observeResource(wasteServiceObserver)
        LogUtils.threadOut("Added ws observer", ColorsOut.CYAN)

        // Force initial update
        CommUtils.delay(200)
        QakContext.getActor("trolley")!!.changed()
    }

    fun cleanupObservers() {
        coapObservers.forEach { it.key.removeObserve(it.value) }
        coapObservers.clear()
        trolleyPosObserver.testend = true
        wasteServiceObserver.testend = true
        println("Cleaned up observers")
    }

    fun waitForHome() {
        println("Waiting for components to be at home")
        while(true) {
            waitForObserverUpdate(5000)
            if (
                wasteServiceObserver.history.last() == "home" &&
                trolleyPosObserver.history.last().contentEquals(intArrayOf(0,0))
            ) {
                break
            }
        }
        wasteServiceObserver.history.let { it.subList(0, it.size-1).clear() }
        trolleyPosObserver.history.let { it.subList(0, it.size-1).clear() }
        println("Observers detected ws&tr at home, can start")
    }
}